### ISSUE TITLE
Parametrize without combinations.

### ISSUE THREAD

**Issue by rvanderwall** [taverntesting/tavern#211]
I am trying to create a parameterized test but don't want the combinations. for example:

```
-parametrize
    key: p1
    vals:
        - a
        - b
-parametrize
    key: p2
        -x
        -y
```
This gives 4 tests, (a,x), (a,y), (b,x), (b,y)but I only want 2 tests (a,x), (b,y)

Another use case is that the pass criteria changes with each parameter.  For example:
```
key: input
vals:
    - 1
    - 3
key: double
vals:
    - 2
    - 6
```
Is that possible? thanks.

**Comment by benhowes** [comment]
@michaelboulton I've had a dig through the docs and tests and can't see any examples of us using parametrize with a list of dictionaries as the `vals`. Should that work? I can add a test/docs if so

**Comment by rvanderwall** [comment]
Each test input may have a different output.  For example, passing in a 2 returns a 4 while passing in a 3 returns a 6.  I want to parameterize tests something like this: (in,out) = [(2, 4), (3,6)] and the test would use {in} on the request and {out} on the response.  Hope that makes sense.

**Comment by michaelboulton** [comment]
I'm assuming you mean like this

```yaml
test_name: parametrise input/output

marks:
  - parametrize:
      key: input
      vals:
        - 1
        - 2
  - parametrize:
      key: output
      vals:
        - 2
        - 4

stages:
  - name: double a number
    request:
      url: "http://example.com/double"
      method: POST
      json:
        value: "{input}"
    response:
      status_code: 200
      body:
        value: "{output}"
```
?

In which case, I think that should work as it is - have you run into any issues with it?

**Comment by rvanderwall** [comment]
I tried that, but I end up with 4 tests, (1,2), (1,4), (2,2), and (2,4).
This may be a different issue, but I cannot parameterize the method.  Looks like it is hard-coded to valid verbs only.  (if it is a different issue, I'll write up a separate request)

**Comment by Birne94** [comment]
While maybe not as clean of a solution, using helper functions you can achieve the desired results:

```yaml
test_name: parametrise input/output

marks:
  - parametrize:
      key: params
      vals:
        - "1|2"
        - "2|4"

stages:
  - name: double a number
    request:
      url: "http://example.com/double"
      method: POST
      json:
        $ext:
          function: utils:create_request
          extra_kwargs:
            param: "{params}"
    response:
      status_code: 200
      body:
        $ext:
          function: utils:verify_response
          extra_kwargs:
            param: "{params}"
```

```python
# utils.py

def create_request(param):
    return {'value': int(param.split('|')[0])}


def verify_response(response, param):
    value = response.json()['value']
    expected = int(param.split('|')[1])

    assert value == expected
```

**Comment by rvanderwall** [comment]
This looks like it may do what I want, but it seems a bit hacky in the sense that a test framework should have a mechanism to specify an input and corresponding output without the need to write parsers.
I'll use this for now, (thanks! i appreciate the solution), but will see if there's a clean way to add this functionality to Tavern itself.
Thanks!

**Comment by benhowes** [comment]
Thanks @Birne94! @rvanderwall we will get something like this added really soon because it is very useful! We probably need to support objects in the vals list in some form

**Comment by michaelboulton** [comment]
> I tried that, but I end up with 4 tests, (1,2), (1,4), (2,2), and (2,4).
> This may be a different issue, but I cannot parameterize the method. Looks like it is hard-coded to valid verbs only. (if it is a different issue, I'll write up a separate request)

Sorry, I completely wasn't thinking when writing that reply, this is the issue that should be fixed by #212 by doing something like this


```yaml
test_name: parametrise input/output

marks:
  - parametrize:
      key: 
        - input
        - output
      vals:
        - [1, 2]
        - [2, 4]

stages:
  - name: double a number
    request:
      url: "http://example.com/double"
      method: POST
      json:
        value: "{input}"
    response:
      status_code: 200
      body:
        value: "{output}"
```

**Comment by rvanderwall** [comment]
This looks like a very good solution.  Do you know when it will be merged and available?
Thanks

**Comment by michaelboulton** [comment]
This should now be available in 0.21.0 , documentation at https://taverntesting.github.io/documentation.html#parametrize

**Comment by 13242084001** [comment]
how to set tavern-merge-ext-function-values? please
